mllib/src/main/scala/org/apache/spark/mllib/linalg/distributed/IndexedRowMatrix.scala

```diff
@@ -113,6 +114,67 @@ class IndexedRowMatrix @Since("1.0.0") (
   }
 
   /**
+    * Converts to BlockMatrix. Creates blocks of `DenseMatrix` with size 1024 x 1024.
+    */
+  def toBlockMatrixDense(): BlockMatrix = {
```

Comment:
I agree that there's no clear default, though if DenseMatrix-backed was the default than people would still be able to get SparseMatrix backed by calling `IndexedRowMatrix.toCoordinateMatrix().toBlockMatrix()`, which is the current implementation of toBlockMatrix anyway. But that's less than ideal, as that's not a particularly good implementation either. 

As for generalizing, if we went ahead and just had one toBlockMatrix method with an argument to specify how it should be backed, it'd be easy enough to take the code that turns the iterator of `(Int, Array[Double])` pairs into a `DenseMatrix` and pull it out into a helper method. Then I could make a corresponding one to make a `SparseMatrix` and just call the appropriate one based on the given argument.